src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs

```diff
@@ -493,6 +495,33 @@ public static string GetMSBuildFilePath(PackageSpec project, RestoreRequest requ
 
                 props.AddRange(GenerateGroupsWithConditions(buildPropsGroup, isMultiTargeting, frameworkConditions));
 
+                // Create an empty PropertyGroup for package properties
+                var packagePathsPropertyGroup = MSBuildRestoreItemGroup.Create("PropertyGroup", Enumerable.Empty<XElement>(), 1000, isMultiTargeting ? frameworkConditions : Enumerable.Empty<string>());
+
+                var projectGraph = targetGraph.Graphs.FirstOrDefault();
+
+                // Distinct union of tool packages and packages with GeneratePathProperty=true
+                var packageIdsToCreatePropertiesFor = packagesWithTools.Union(projectGraph.Item.Data.Dependencies.Where(i => i.GeneratePathProperty).Select(i => i.Name)).Distinct(StringComparer.OrdinalIgnoreCase);
+
+                // Find the packages with matching IDs in the list of sorted packages, filtering out ones that there was no match for or that don't exist
+                var packagePathProperties = packageIdsToCreatePropertiesFor
+                    // Cast sortedPackages as nullable
+                    .Select(packageId => sortedPackages.Cast<KeyValuePair<LockFileTargetLibrary, Lazy<LocalPackageSourceInfo>>?>().FirstOrDefault(pkg => pkg != null && pkg.Value.Key.Name.Equals(packageId, StringComparison.OrdinalIgnoreCase)))
```

Comment:
You don't really need to use the LockFileTargetLibrary at all here. 

I think you can probably use the localPackage here, pkg.Value.Value.Key.Name.
Then you do away with the key in the select clause below.

Reply:
Maybe even avoid the cast by preselecting the localPackageInfo only. 

Even if it's 2 clauses it's more readable than it currently is.

Reply:
Synced up offline with @jeffkl 
```
                var localPackages = sortedPackages.Select(e => e.Value);
                // Find the packages with matching IDs in the list of sorted packages, filtering out ones that there was no match for or that don't exist
                var packagePathProperties = localPackages
                    .Where(pkg => packageIdsToCreatePropertiesFor.Contains(pkg.Value.Package.Id) && pkg.Exists())
                    .Select(pkg => pkg.Value.Package)
                    // Get the property
                    .Select(i => GeneratePackagePathProperty(i));

```